src/peer_sourcehash.cc

```diff
@@ -174,17 +176,15 @@ peerSourceHashSelectParent(HttpRequest * request)
         score = combined_hash * tp->sourcehash.load_multiplier;
         debugs(39, 3, "peerSourceHashSelectParent: " << tp->name << " combined_hash " << combined_hash  <<
                " score " << std::setprecision(0) << score);
-
-        if ((score > high_score) && peerHTTPOkay(tp, request)) {
-            p = tp;
-            high_score = score;
-        }
+        if (peerHTTPOkay(tp, request))
+            sortedPeers.push_back(std::pair<int, CachePeer *>(score, tp));
     }
 
-    if (p)
-        debugs(39, 2, "peerSourceHashSelectParent: selected " << p->name);
-
-    return p;
+    std::sort(sortedPeers.begin(), sortedPeers.end());
+    for (auto it = sortedPeers.rbegin(); it!= sortedPeers.rend(); ++it) {
```

Comment:
True. This is a major bug of the patch.

> In other words, after the first match, the rest of the group is discarded. I think this can be done cheaply by using a new integer FwdServer::selectionGroup that PeerSelector would assign and consult as needed.

If a peer rejected because of an acl check or because it is down, it should discarded from all other groups however. Or keep a list of rejected peers and do not try to reuse them. Or maybe modify the list while it is checked.

Reply:
You can remove the disqualified peer from the remaining candidate pool when that peer is disqualified. This way, you do not need to keep more state -- a new selectionGroup field is enough.

BTW, when a peer is _qualified_, then the code should remove the remaining same-group peers from the remaining candidate pool. Thus, regardless of whether the peer is qualified or disqualified, the code needs to remove certain peers (if any).